tools/wptrunner/wptrunner/executors/executorcontentshell.py

```diff
@@ -150,8 +150,33 @@ def read_errors(self):
         return result
 
 
+class ContentShellBasePart(ProtocolPart):
+    """This protocol part provides functionality common to all executors.
+
+    In particular, this protocol part implements `wait()`, which, when
+    `--pause-after-test` is enabled, test runners block on until the next test
+    should run.
+    """
+    name = "base"
+
+    def __init__(self, parent):
+        super().__init__(parent)
+        self.io_stopped = parent.browser.io_stopped
+
+    def wait(self):
+        # This worker is unpaused when the browser window is closed, which this
```

Comment:
Does this mean content shell exits? If that is the case, did we restart content shell properly for next test?

Reply:
Yes, `self.io_stopped.wait()` unblocks once `content_shell` stdout/stderr reaches EOF:

https://github.com/web-platform-tests/wpt/blob/e432c73c6a769442839648d31dd2b079fb267a27/tools/wptrunner/wptrunner/browsers/content_shell.py#L209

This is similar to the default WebDriver implementation of `wait()`, which blocks until the tab closes (`NoSuchWindowException`):

https://github.com/web-platform-tests/wpt/blob/e432c73c6a769442839648d31dd2b079fb267a27/tools/wptrunner/wptrunner/executors/executorwebdriver.py#L78-L96

The test runner is guaranteed to restart content shell because `wait()` returns `rerun=False <=> restart=True` identically:

https://github.com/web-platform-tests/wpt/blob/e432c73c6a769442839648d31dd2b079fb267a27/tools/wptrunner/wptrunner/testrunner.py#L756

(Aside: This change is necessary but not yet sufficient to get headful debugging to work with `content_shell`, which I have not yet figured out.)

Reply:
Thanks for the explanation.